bpo-36044: Reduce number of unit tests run for PGO build.

nascheme opened the pull request:
This speeds up the PGO profile generation by a factor of about 15X.
Running the full unit test suite is slow.  This may result in a
slightly less optimized build since not as many code branches will
be executed.  If you are willing to wait for the slower build, the
old behavior can be enabled as follows:

    ./configure [...] --with-profile-task='-m test --pgo'

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-36044](https://bugs.python.org/issue36044) -->
https://bugs.python.org/issue36044
<!-- /issue-number -->

nascheme committed:
Reduce number of unit tests run for PGO build.

This speeds up the PGO profile generation by a factor of about 15X.
Running the full unit test suite is slow.  This may result in a
slightly less optimized build since not as many code branches will
be executed.  If you are willing to wait for the slower build, the
old behavior can be enabled as follows:

    ./configure [...] --with-profile-task='-m test --pgo'

nascheme committed:
Add NEWS.

nascheme committed:
Move NEWS entry under "Build"

nascheme committed:
Move list of PGO tests from Makefile to regrtest.

Add --pgo-extended option to regrtest.

vstinner commented on Lib/test/libregrtest/cmdline.py:
```diff
@@ -264,7 +264,9 @@ def _create_parser():
     group.add_argument('-P', '--pgo', dest='pgo', action='store_true',
-                       help='enable Profile Guided Optimization training')
+                       help='enable Profile Guided Optimization (PGO) training')
+    group.add_argument('--pgo-extended', action='store_true',
```
I don't understand what "extended" means. I suggest to rename the option to "--pgo-all-tests" and rewrite the help to explain that it runs all tests by default.

nascheme replied:
Think list.extend().  I.e. run more tests.  I didn't want to use --pgo-all because we don't actually run all the tests.  Some are excluded using the `support.PGO` flag.  However, I don't feel strongly so if you really like --pgo-all, I will change it.

nascheme replied:
Hmm, --pgo-harder?  :smile:

vstinner commented on Lib/test/libregrtest/main.py:
```diff
@@ -21,6 +21,55 @@
+# chosen based on the following criteria: either they exercise a commonly used
+# C extension module or type, or they run some relatively typical Python code.
+# Long running tests should be avoided because the PGO instrumented executable
+# runs slowly.
```
I would prefer to keep main.py small. Would you mind to create libregrtest/pgo.py for this list please?

nascheme replied:
Ok.

nascheme committed:
Move default list of PGO tests to own module.

gpshead commented on Lib/test/libregrtest/cmdline.py:
```diff
@@ -264,7 +264,9 @@ def _create_parser():
-                       help='enable Profile Guided Optimization training')
+                       help='enable Profile Guided Optimization (PGO) training')
+    group.add_argument('--pgo-extended', action='store_true',
+                       help='enable extended PGO training (slower training)')
```
I'd add to the description of this that "User are encouraged to benchmark the interpreter resulting from a PROFILE_TASK with this flag to decide if it is meaningfully faster than those produced using --pgo for their use cases".

I don't care what the flag name is, --pgo-extended is fine by me.  as is --pgo-very-long-training-run (a flag who's length matches the build time seems... appropriate). ;)

gpshead committed:
update the news wording a bit.

more detail and a performance disclaimer.

gpshead reviewed (approved):
i updated the news entry wording a bit.

nascheme committed:
Set PROFILE_TASK from env or configure args.

The configure flags --with-something are intended for enabling
optional external software.  Just use a variable instead.
